src/ductile/lifecycle.clj

```diff
@@ -0,0 +1,358 @@
+
+(s/defschema ILMPhase
+  "ILM phase structure"
+  (s/conditional
+   #(contains? % :actions)
+   {(s/optional-key :min_age) s/Str
+    :actions {s/Keyword s/Any}}
+   :else
+   {s/Keyword s/Any}))
+
+(s/defschema ILMPolicy
+  "Elasticsearch ILM policy structure"
+  {:phases {s/Keyword ILMPhase}})
+
+(s/defschema ISMTransition
+  "OpenSearch ISM transition structure"
+  {(s/optional-key :state_name) s/Str
+   (s/optional-key :conditions) {s/Keyword s/Any}})
+
+(s/defschema ISMAction
+  "OpenSearch ISM action structure"
+  (s/conditional
+   map? {s/Keyword s/Any}
+   :else s/Any))
+
+(s/defschema ISMState
+  "OpenSearch ISM state structure"
+  {:name s/Str
+   :actions [ISMAction]
+   (s/optional-key :transitions) [ISMTransition]})
+
+(s/defschema ISMPolicy
+  "OpenSearch ISM policy structure"
+  {:states [ISMState]
+   (s/optional-key :description) s/Str
+   (s/optional-key :default_state) s/Str
+   (s/optional-key :schema_version) s/Int
+   (s/optional-key :ism_template) [{s/Keyword s/Any}]})
+
+;; ============================================================================
+;; ILM Action Transformation
+;; ============================================================================
+
+(defn- transform-rollover-action
+  "Transform ILM rollover action to ISM format"
+  [rollover-config]
+  {:rollover
+   (cond-> {}
+     (:max_age rollover-config)
+     (assoc :min_index_age (:max_age rollover-config))
+
+     (:max_docs rollover-config)
+     (assoc :min_doc_count (:max_docs rollover-config))
+
+     (:max_size rollover-config)
+     (assoc :min_size (:max_size rollover-config)))})
+
+(defn- transform-delete-action
+  "Transform ILM delete action to ISM format"
+  [_delete-config]
+  {:delete {}})
+
+(defn- transform-readonly-action
+  "Transform ILM readonly action to ISM format"
+  [_readonly-config]
+  {:read_only {}})
+
+(defn- transform-shrink-action
+  "Transform ILM shrink action to ISM format"
+  [shrink-config]
+  {:shrink
+   (cond-> {}
+     (:number_of_shards shrink-config)
+     (assoc :num_new_shards (:number_of_shards shrink-config)))})
+
+(defn- transform-force-merge-action
+  "Transform ILM force_merge action to ISM format"
+  [force-merge-config]
+  {:force_merge
+   (cond-> {}
+     (:max_num_segments force-merge-config)
+     (assoc :max_num_segments (:max_num_segments force-merge-config)))})
+
+(defn- transform-ilm-action
+  "Transform a single ILM action to ISM format"
+  [[action-name action-config]]
+  (case action-name
+    :rollover (transform-rollover-action action-config)
+    :delete (transform-delete-action action-config)
+    :readonly (transform-readonly-action action-config)
+    :shrink (transform-shrink-action action-config)
+    :force_merge (transform-force-merge-action action-config)
+    ;; Unsupported actions are logged but not transformed
+    (do
+      (when-not (#{:set_priority :allocate :migrate} action-name)
+        (println (str "Warning: Unsupported ILM action: " action-name)))
```

Comment:
Should use clojure.tools.logging/warn instead of println for consistency with the
rest of the codebase.